hw/ip/otbn/util/shared/testcase.py

```diff
@@ -0,0 +1,142 @@
+        return regs
+
+    @staticmethod
+    def _from_dict(data: Dict[str, Any],
+                   symbols: Dict[str, int]) -> 'StateExpectations':
+        unknown_keys = set(data.keys()) - {'dmem', 'regs'}
+        if len(unknown_keys):
+            raise ValueError(f'Got unknown keys in testcase input/output: {unknown_keys}')
+
+        dmem_dict = data.get('dmem', {})
+        regs_dict = data.get('regs', {})
+        return StateExpectations(
+            dmem=StateExpectations._dmem_from_dict(dmem_dict, symbols),
+            regs=StateExpectations._regs_from_dict(regs_dict, symbols)
+        )
+
+
+@dataclass
+class OtbnTestCase:
+    entrypoint: int
+    input: StateExpectations
+    output: StateExpectations
+
+    @staticmethod
+    def from_hjson(hjson_str: str,
+                   symbols: Dict[str, int]) -> 'OtbnTestCase':
+        """
+        Parse a testcase HJSON string into a OtbnTestCase object.
+
+        The HJSON schema is defined as follows:
+        ```
+            {
+                entrypoint: str = "0"  // Address or symbol of entrypoint
+                input: StateExpectations
+                output: StateExpectations
+            }
+        ```
+
+        The StateExpectations schema is defined as follows:
+        ```
+            {
+                dmem: {
+                    // Value is hex string (e.g. "0xdeadbeef") or symbol (e.g. "symbol").
+                    //
+                    // The variable length is inferred from the value's length or
+                    // is 4 bytes if the value is a symbol.
+                    <label>: <value>
+                },
+                regs: {
+                    // Value is int, hex string (e.g. "0xdeadbeef"), or symbol (e.g. "symbol").
+                    <reg_name>: <value>
+                }
+            }
+        ```
+
+        Args:
+            hjson_str: The HJSON string to parse.
+            symbols: A dict from symbol name (str) to value (int). This should contain
+                     all symbols from the program being tested.
+
+        Returns:
+            A OtbnTestCase object representing the parsed testcase.
+        """
+        data: Dict[str, Any] = hjson.loads(hjson_str)
+
+        unknown_keys = set(data.keys()) - {'entrypoint', 'input', 'output'}
+        if len(unknown_keys):
+            raise ValueError(f'Got unknown keys in testcase: {unknown_keys}')
+
+        entrypoint = data.get('entrypoint', 0)
+        if isinstance(entrypoint, int):
+            entrypoint_addr = entrypoint
+        elif _is_numeric_str(entrypoint):
+            entrypoint_addr = int(entrypoint, 0)
+        else:
+            entrypoint_addr = _get_symbol(symbols, entrypoint)
+
+        input_data = data.get('input', {})
+        output_data = data.get('output', {})
+
+        return OtbnTestCase(
+            entrypoint=entrypoint_addr,
+            input=StateExpectations._from_dict(input_data, symbols),
```

Comment:
I'd probably inline the `data.get` calls here as well (getting rid of the `input_data` and `output_data` variables).